src/cls/rbd/cls_rbd_types.h

```diff
@@ -148,12 +183,14 @@ struct MirrorImage {
   MirrorImage() {
   }
   MirrorImage(MirrorImageMode mode, const std::string &global_image_id,
-              MirrorImageState state)
-    : mode(mode), global_image_id(global_image_id), state(state) {
+              const GroupSpec &group_spec, MirrorImageState state)
```

Comment:
For discussion - I think the group information in MirrorImage should be was mirror specific , ie, global_group_id instead of the group_id. The mirroring data structures usually work with global ids which makes it easier in case of resyncs. The GroupSpec makes sense in a non-mirror group snapshot scenario.

Reply:
Note: this would imply that an image in a group cannot be mirrored individually as is possible now.

Reply:
> > Note: The existing code allows an image in a group to be mirrored independently by using an empty group spec. The rbd_group_ref information is not mirrored so the image is not part of a group on the remote cluster.
> 
> This shouldn't be allowed, right?

This is used in the case of mirroring an image which is not part of a group.

Reply:
It is part of the MirrorImage struct in the CG PR changes. For the mirror information it makes more sense to use global information.

Reply:
``
struct MirrorImage {
  MirrorImage() {
  }
  MirrorImage(MirrorImageMode mode, const std::string &global_image_id,
              const GroupSpec &group_spec, MirrorImageState state)
    : mode(mode), global_image_id(global_image_id), group_spec(group_spec),
      state(state) {
  }

  MirrorImageMode mode = MIRROR_IMAGE_MODE_JOURNAL;
  std::string global_image_id;
  GroupSpec group_spec;
  MirrorImageState state = MIRROR_IMAGE_STATE_DISABLING;
``


See `MirrorGroupImageSpec` in `MirrorGroup` which does only use global ids. The same approach should be used here **iff** we do not intend to support mirroring an individual image in a group.

Reply:
> The same approach should be used here **iff** we do not intend to support mirroring an individual image in a group.

According to prior discussions, we don't.

But I wonder if `MirrorImage` needs to change at all.  The group spec is available in the image header in the form of `rbd_group_ref` key, while `global_group_id` is available in `MirrorGroup`.  Neither seems to add any additional information if stored here.

Reply:
> > Note: The existing code allows an image in a group to be mirrored independently by using an empty group spec. The rbd_group_ref information is not mirrored so the image is not part of a group on the remote cluster.
> 
> Just curious, is it exposed all the way up to the user or just possible if an empty ("invalid") group spec somehow gets produced here? Either way, I think this ability should be disabled.

I didn't understand this point. Can you elaborate?

Reply:
The GroupSpec in the MirrorImage is also used to filter out images that are mirrored as part of a group.
https://github.com/ceph/ceph/pull/53793/commits/1cc4ade7119f189d97636fbf84711ba84b47b477

@pkalever , what would be the impact of not having this?

Reply:
> > Just curious, is it exposed all the way up to the user or just possible if an empty ("invalid") group spec somehow gets produced here? Either way, I think this ability should be disabled.
> 
> I didn't understand this point. Can you elaborate?

I was wondering whether it's possible for the user to enable mirroring on an image via regular `rbd mirror image enable` command.  After the discussion in yesterday's meeting, I think the answer is yes and we all agree that this should be plugged.